### PR TITLE
Use conda compilers

### PR DESCRIPTION
**Pull request by Ethyling**
This PR enables the usage of conda compilers to build conda packages. This is required to use `mambabuild`

### CODE REVIEW

**Comment by Ethyling** [comment]
rerun tests

**Comment by Ethyling** [comment]
rerun tests

**Comment by jakirkham** [comment]
In [the build log]( https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=11.0,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/8007/timestamps/?time=HH:mm:ss&timeZone=GMT-7&appendLog&locale=en_US ) am seeing the following

```
10:06:53  $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -isystem $PREFIX/include -fPIC -isystem $PREFIX/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/cudf-22.06.00a -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I/usr/local/cuda/include -DNDEBUG -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -I/usr/local/cuda/include -fPIC -O3 -Wno-unknown-pragmas -I$PREFIX/lib/python3.8/site-packages/cuda -I$SRC_DIR/cpp/include/cudf -I$SRC_DIR/cpp/include -I$SRC_DIR/cpp/build/include -I$SRC_DIR/cpp/build/_deps/libcudacxx-src/include -I$SRC_DIR/cpp/build/_deps/dlpack-src/include -I$PREFIX/include/rapids/libcudacxx -I$PREFIX/include -I$PREFIX/lib/python3.8/site-packages/numpy/core/include -I$PREFIX/lib/python3.8/site-packages/pyarrow/include -I/usr/local/cuda/include -I$PREFIX/include/python3.8 -c cudf/_lib/column.cpp -o build/temp.linux-x86_64-cpython-38/cudf/_lib/column.o -std=c++17
```

In particular am seeing

```
-I$SRC_DIR/cpp/build/_deps/libcudacxx-src/include ... -I$PREFIX/include/rapids/libcudacxx
```

Wonder if including both of these is causing issues

**Comment by Ethyling** [comment]
> In [the build log](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/cudf/job/prb/job/cudf-gpu-test/CUDA=11.0,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/8007/timestamps/?time=HH:mm:ss&timeZone=GMT-7&appendLog&locale=en_US) am seeing the following
> 
> ```
> 10:06:53  $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -isystem $PREFIX/include -fPIC -isystem $PREFIX/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/cudf-22.06.00a -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I/usr/local/cuda/include -DNDEBUG -D_FORTIFY_SOURCE=2 -isystem $PREFIX/include -I/usr/local/cuda/include -fPIC -O3 -Wno-unknown-pragmas -I$PREFIX/lib/python3.8/site-packages/cuda -I$SRC_DIR/cpp/include/cudf -I$SRC_DIR/cpp/include -I$SRC_DIR/cpp/build/include -I$SRC_DIR/cpp/build/_deps/libcudacxx-src/include -I$SRC_DIR/cpp/build/_deps/dlpack-src/include -I$PREFIX/include/rapids/libcudacxx -I$PREFIX/include -I$PREFIX/lib/python3.8/site-packages/numpy/core/include -I$PREFIX/lib/python3.8/site-packages/pyarrow/include -I/usr/local/cuda/include -I$PREFIX/include/python3.8 -c cudf/_lib/column.cpp -o build/temp.linux-x86_64-cpython-38/cudf/_lib/column.o -std=c++17
> ```
> 
> In particular am seeing
> 
> ```
> -I$SRC_DIR/cpp/build/_deps/libcudacxx-src/include ... -I$PREFIX/include/rapids/libcudacxx
> ```
> 
> Wonder if including both of these is causing issues

@jakirkham I think this is coming from https://github.com/rapidsai/cudf/blob/2aaa863ee78f4cdc8d7a0180abe82bdad4cd7923/python/cudf/setup.py#L207 and https://github.com/rapidsai/cudf/blob/2aaa863ee78f4cdc8d7a0180abe82bdad4cd7923/python/cudf/setup.py#L209-L212. I'm not sure what's the idea behind this and if it's causing any issue

**Comment by Ethyling** [comment]
rerun tests

**Comment by jakirkham** [comment]
Yeah it looks like the former is leftover and probably can be dropped. The latter is being used with `rapids_cpm` ( https://github.com/rapidsai/cudf/pull/9539 ). Though if it isn't causing build issues here, maybe we can revisit it separately.

**Comment by Ethyling** [comment]
rerun tests

File: conda/recipes/dask-cudf/meta.yaml
Comment (bdice) [inline conda/recipes/dask-cudf/meta.yaml:30]: It might be possible to remove CUDA compiler requirement -- I don't think there are any CUDA-compiled pieces in dask-cudf. The only catch is that `setup.py` checks the installed CUDA toolkit to determine which version of cupy to use from pip! That doesn't apply in the conda builds, though, because matching cupy and CUDA toolkit versions are handled by conda. 😕 https://github.com/rapidsai/cudf/blob/e58d0490472467c724cd9d9e5ec1c6f6efa4beae/python/dask_cudf/setup.py#L68-L70

See also: https://pypi.org/project/cupy/ for list of prebuilt wheels by CUDA version.

A few notes on dependencies:
- `dask_cudf` should probably list `cupy, numpy, pandas, fsspec, pyarrow` as runtime requirements and not just get them transitively from the dependency on `cudf`. Those are listed in `setup.py` except for `pyarrow`, which should probably be added there as well.
- I also see dependencies on `toolz` (listed as `import tlz`) which seems to be a dependency of dask that we're getting transitively. https://github.com/dask/dask/issues/1849#issuecomment-806715040

This could all be handled in a separate PR, I just wanted to make some notes for the moment.

File: cpp/include/cudf/fixed_point/temporary.hpp
Comment (bdice) [inline cpp/include/cudf/fixed_point/temporary.hpp:21]: Why was this added? I looked at the file but I don't see what requires this.

**Comment by Ethyling** [comment]
rerun tests

File: conda/recipes/libcudf/meta.yaml
Comment (ajschmidt8) [inline conda/recipes/libcudf/meta.yaml:23]: Is this change intended to stay?

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Review by galipremsagar — APPROVED** [review]

**Comment by Ethyling** [comment]
@gpucibot merge

File: conda/recipes/custreamz/meta.yaml
Comment (vyasr) [inline conda/recipes/custreamz/meta.yaml:30]: Does custreamz not need cudf? It's just a transitive dependency via cudf_kafka? I realize this isn't a change, but since the cudf/meta.yaml does look to have been changed meaningful I want to double-check this one as well.

File: conda/recipes/cudf/meta.yaml
Comment (bdice) [inline conda/recipes/cudf/meta.yaml:31]: Do we mean the following? I don't know if `sysroot_linux-64` is only supposed to be installed on `x86_64`. See [preprocessing selectors](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preprocessing-selectors). I believe that Linux ARM systems match both `linux64` and `aarch64`.
```suggestion
    - sysroot_linux-64 {{ sysroot_version }}  # [linux64 and x86_64]
    - sysroot_linux-aarch64 {{ sysroot_version }}  # [linux64 and aarch64]
```

File: conda/recipes/cudf/conda_build_config.yaml
Comment (vyasr) [reply conda/recipes/cudf/conda_build_config.yaml:5]: Confirmed offline that this is just matching our preexisting compilation.

File: cpp/libcudf_kafka/CMakeLists.txt
Comment (vyasr) [inline cpp/libcudf_kafka/CMakeLists.txt:36]: Why was this added? I don't see any new targets here.

File: cpp/include/cudf/fixed_point/temporary.hpp
Comment (vyasr) [reply cpp/include/cudf/fixed_point/temporary.hpp:21]: This looks like an issue where libcudacxx also defines `__forceinline__` and it's clashing with the default cuda definition. My guess is that the libcudacxx version has ifdef guards, while the cuda runtime does not. I am not sure why that changes with this though. Would be helpful to get @jrhemstad's opinion, but I don't think this needs to be a blocker for this PR since we need to unblock our CI.

File: conda/recipes/cudf/conda_build_config.yaml
Comment (Ethyling) [reply conda/recipes/cudf/conda_build_config.yaml:5]: We are keeping the same version used before conda compilers

File: conda/recipes/custreamz/meta.yaml
Comment (Ethyling) [reply conda/recipes/custreamz/meta.yaml:30]: Yes, `cudf_kafka` depends on `cudf`, so `custreamz` depends on `cudf` too

Comment (bdice) [reply conda/recipes/custreamz/meta.yaml:30]: On that note, why does `cudf` match to `{{ version }}.*` but `cudf_kafka` only matches `{{ version }}`?
```suggestion
    - cudf {{ version }}.*
    - cudf_kafka {{ version }}
```

File: cpp/include/cudf/fixed_point/temporary.hpp
Comment (jrhemstad) [reply cpp/include/cudf/fixed_point/temporary.hpp:21]: Given this PR is mucking with the compilers, I suspect something bad is happening.

File: conda/recipes/custreamz/meta.yaml
Comment (Ethyling) [reply conda/recipes/custreamz/meta.yaml:30]: I tried to unify the recipes without changing everything, but all the recipes should match `{{ version }}.*`, but I think we should do this in another PR. That's on my TODO list

File: conda/recipes/cudf/meta.yaml
Comment (jakirkham) [reply conda/recipes/cudf/meta.yaml:31]: It is

Comment (bdice) [reply conda/recipes/cudf/meta.yaml:31]: I don't think that's the case, but I'm not able to find the conda source code that handles this. I think this pattern is misused by a lot of recipes and feedstocks which default to `linux64` and overwrite the value on `aarch64` and other platforms, if my understanding of selector logic is correct. However, this isn't a case where we're overriding a single value, but adding multiple values to the same list. If the recipe works as-is, we don't need to block on this, but I would like to clarify my understanding since "`linux64` is for `x86_64` only" does not align with the docs I linked above. cc: @jakirkham in case you have insight here.

Comment (bdice) [reply conda/recipes/cudf/meta.yaml:31]: Sorry @jakirkham I missed your comment. I see it now that I refreshed. All good.

Comment (jakirkham) [reply conda/recipes/cudf/meta.yaml:31]: That said, we can simplify this like so

```suggestion
    - sysroot_{{ target_platform }} {{ sysroot_version }}
```

File: conda/recipes/cudf_kafka/meta.yaml
Comment (jakirkham) [inline conda/recipes/cudf_kafka/meta.yaml:29]:
```suggestion
    - sysroot_{{ target_platform }} {{ sysroot_version }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:35]:
```suggestion
    - sysroot_{{ target_platform }} {{ sysroot_version }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:55]: Can use `compiler` syntax here

```suggestion
      ignore_run_exports_from:
        - {{ compiler('cuda') }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:297]:
```suggestion
      ignore_run_exports_from:
        - {{ compiler('cuda') }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:321]:
```suggestion
      ignore_run_exports_from:
        - {{ compiler('cuda') }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:328]:
```suggestion
        - sysroot_{{ target_platform }} {{ sysroot_version }}
```

File: conda/recipes/libcudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/libcudf/meta.yaml:347]:
```suggestion
      ignore_run_exports_from:
        - {{ compiler('cuda') }}
```

**Review by jakirkham — COMMENTED** [review]
Thanks Jordan! 😄

Generally this looks great. Had a few suggested simplifications below

File: cpp/include/cudf/fixed_point/temporary.hpp
Comment (vyasr) [reply cpp/include/cudf/fixed_point/temporary.hpp]: It looks like this still fails without the explicit include of `cudf_runtime.h`. Perhaps that means something is still off with the conda nvcc compiler and its sysroot paths?

File: conda/recipes/cudf/meta.yaml
Comment (jakirkham) [reply conda/recipes/cudf/meta.yaml:31]: Also submitted PR ( https://github.com/conda/conda-build/pull/4476 ) to clarify the docs

**Comment by Ethyling** [comment]
@jakirkham PR updated

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks Jordan! 🙏 LGTM 😄

File: cpp/include/cudf/fixed_point/temporary.hpp
Comment (Ethyling) [reply cpp/include/cudf/fixed_point/temporary.hpp]: Fixed

**Comment by vyasr** [comment]
It looks like the most recent release of [fsspec](https://anaconda.org/conda-forge/fsspec/) broke a few of our tests while our CI was offline. The issue fsspec/filesystem_spec#961, which changes the behavior of `fsspec.core.get_fs_token_path` to correctly handle all globs. Previously, that function only handled globs containing an asterisk `*`, but that PR fixed the function to also handle `?` and `[]`. The problem is that we have tests of `cudf.read_json` with strings containing `[]` (because they contain lists), and when those globs read [this line](https://github.com/rapidsai/cudf/blob/branch-22.06/python/cudf/cudf/utils/ioutils.py#L1215) in the code they correspond to no preexisting files and so instead of getting a path to a nonexistent file like we would have before, we instead get an empty list of paths. That in turn leads to us hitting [this error case](https://github.com/rapidsai/cudf/blob/branch-22.06/python/cudf/cudf/utils/ioutils.py#L1375).

Specifically, here is the difference:
```
# The new behavior
>>> fsspec.get_fs_token_paths(['[true]'], mode='rb')
(<fsspec.implementations.local.LocalFileSystem object at 0x7f93c7d58040>, '00e91f08872e08eaa06e0006dfc00ab6', [])

# The old behavior. Note that the third element of the tuple was not an empty string, but instead a path.
>>> fsspec.get_fs_token_paths(['[]'], mode='rb')
(<fsspec.implementations.local.LocalFileSystem object at 0x7fa94db5feb0>, '839c5f114235189ec7ed1e0a405abf70', ['/home/vyasr/[]'])

```

@rjzamora how do you recommend that we fix this? The change to fsspec seems correct, so it's not a bug for us to report upstream but rather something that we need to fix in our usage. I think we probably want to change this error check because using the fact that no paths were returned as an error condition presupposes that the input was a a valid path when it could be an arbitrary JSON string, but I don't have all the context here. Perhaps there is a better solution that addresses problems that I'm not seeing.

**Comment by galipremsagar** [comment]
> It looks like the most recent release of [fsspec](https://anaconda.org/conda-forge/fsspec/) broke a few of our tests while our CI was offline. The issue is fsspec/filesystem_spec#961, which changes the behavior of `fsspec.core.get_fs_token_path` to correctly handle all globs. Previously, that function only handled globs containing an asterisk `*`, but that PR fixed the function to also handle `?` and `[]`. The problem is that we have tests of `cudf.read_json` with strings containing `[]` (because they contain lists), and when those globs read [this line](https://github.com/rapidsai/cudf/blob/branch-22.06/python/cudf/cudf/utils/ioutils.py#L1215) in the code they correspond to no preexisting files and so instead of getting a path to a nonexistent file like we would have before, we instead get an empty list of paths. That in turn leads to us hitting [this error case](https://github.com/rapidsai/cudf/blob/branch-22.06/python/cudf/cudf/utils/ioutils.py#L1375).
> 
> 
> 
> Specifically, here is the difference:
> 
> ```
> 
> # The new behavior
> 
> >>> fsspec.get_fs_token_paths(['[true]'], mode='rb')
> 
> (<fsspec.implementations.local.LocalFileSystem object at 0x7f93c7d58040>, '00e91f08872e08eaa06e0006dfc00ab6', [])
> 
> 
> 
> # The old behavior. Note that the third element of the tuple was not an empty string, but instead a path.
> 
> >>> fsspec.get_fs_token_paths(['[]'], mode='rb')
> 
> (<fsspec.implementations.local.LocalFileSystem object at 0x7fa94db5feb0>, '839c5f114235189ec7ed1e0a405abf70', ['/home/vyasr/[]'])
> 
> 
> 
> ```
> 
> 
> 
> @rjzamora how do you recommend that we fix this? The change to fsspec seems correct, so it's not a bug for us to report upstream but rather something that we need to fix in our usage. I think we probably want to change this error check because using the fact that no paths were returned as an error condition presupposes that the input was a a valid path when it could be an arbitrary JSON string, but I don't have all the context here. Perhaps there is a better solution that addresses problems that I'm not seeing.

Yea, I ran into this yesterday. I've worked on a fix. Will open the PR over the weekend.

**Comment by vyasr** [comment]
Thanks @galipremsagar! No rush, I doubt this will seriously block anyone before Monday. I just wanted to diagnose this ASAP.

**Comment by galipremsagar** [comment]
Opened a PR to address the json pytest failures here: https://github.com/rapidsai/cudf/pull/10924

@Ethyling could you admin merge this(10275) PR since everyone has approved it?

**Pull request merged**
